File: algorithms/Fraction to Recurring Decimal/Fraction to Recurring Decimal.cpp
class Solution {
public:
    string fractionToDecimal(int numerator, int denominator) {
        unordered_map<int, int> umap;
        string result = "";
        if ((double) numerator / (double) denominator < 0) result.push_back('-');
        long long l_numerator = numerator > 0 ? numerator : -(long long) numerator;
        long long l_denominator = denominator > 0 ? denominator : -(long long) denominator;
        long long quotient = l_numerator / l_denominator;
        long long remainder = l_numerator % l_denominator;
        result.append(to_string(quotient));
        if (remainder == 0) return result;
        result.push_back('.');
        int position = result.size();
        umap[remainder] = position++;
        while (remainder != 0) {
            l_numerator = remainder * 10;
            quotient = l_numerator / l_denominator;
            remainder = l_numerator % l_denominator;
            char digit = '0' + quotient;
            result.push_back(digit);
            if (umap.find(remainder) != umap.end()) {
                result.insert(umap[remainder], 1, '(');
                result.push_back(')');
                return result;
            }
            umap[remainder] = position++;
        }
        return result;
    }
};

File: algorithms/Fraction to Recurring Decimal/Fraction to Recurring Decimal.py
from collections import defaultdict
class Solution:
    def fractionToDecimal(self, numerator: int, denominator: int) -> str:
        sign = "" if numerator*denominator >= 0 else "-"
        numerator, denominator = abs(numerator), abs(denominator)
        a = numerator//denominator
        numerator %= denominator
        if not numerator: return sign+str(a)
        fractions = []
        index = defaultdict(int)
        while 10*numerator not in index:
            numerator *= 10
            index[numerator] = len(fractions)
            fractions.append(str(numerator//denominator))
            numerator %= denominator
        i = index[10*numerator]
        return sign+str(a)+"."+"".join(fractions[:i])+"("+"".join(fractions[i:])+")" if numerator else sign+str(a)+"."+"".join(fractions[:i])